static/js/blueslip_stacktrace.ts

```diff
@@ -0,0 +1,113 @@
+        full_path = show_path = data.fileName.slice(idx);
+    } else {
+        full_path = data.fileName.split('/').slice(-3).join('/');
+        show_path = full_path.slice(full_path.lastIndexOf('/') + 1);
+    }
+
+    // Don't show the full scope in function name
+    let full_fn_name;
+    try {
+        full_fn_name = /(?:[a-z_]+\.[jt]s(?:\?\w+)?\.)(.+)/.exec(data.functionName)[1];
+    } catch {
+        full_fn_name = data.functionName;
+    }
+    const fn_name = full_fn_name.split('.');
+
+    fn_name[fn_name.length - 1] = '<span class="function-name">'
+        + fn_name[fn_name.length - 1].replace('<', '&lt;').replace('>', '&gt;')
+        + '</span>';
+
+    return {
+        full_path: full_path,
+        show_path: show_path,
+        line_number: data.lineNumber,
+        function_name: fn_name.join('.'),
+    };
+}
+
+const stack_trace_gps = new StackTraceGPS();
+
+async function beautify_stackframe(stackframe: StackFrame): Promise<CleanStackFrame> {
+    return stack_trace_gps
+        .getMappedLocation(stackframe)
+        .then(clean_stackframe_data);
+}
+
+export function display_stacktrace(error: string, stack: string): void {
+    const ex = new Error();
+    ex.stack = stack;
+    const stacktrace: Promise<CleanStackFrame>[] =
+        ErrorStackParser.parse(ex).map(beautify_stackframe);
+
+    // The stacktrace-gps library uses Promise in their APIs for fetching
+    // source maps from the server and the underlying mozilla/source-map
+    // library is compiled to wasm. We assume the developer's browser is recent
+    // enough to support both of these.
+    Promise.all(stacktrace).then(stackframes => {
+        const template = _.template($('.stacktrace #stackframe-template').html());
+        return stackframes.map(template).join('');
+    }).then(rendered_stackframes => {
+        $('.stacktrace .stacktrace-content').html(rendered_stackframes);
+        $('.stacktrace .message').html('<b>Exception</b> ' + error);
+        $('.stacktrace').addClass('show');
+    });
+}
+
+function expand_code_context(target: JQuery): void {
+    if (target.css('maxHeight') !== '0px') {
+        target.css('maxHeight', '0px');
+    } else {
+        target.css('maxHeight', target[0].scrollHeight + 'px');
+    }
+    target.toggleClass('show');
+}
+
+export function display_code_context(
+    full_path: string,
+    line_number: number,
+    context_wrapper: JQuery
+): void {
+    if (context_wrapper.children().first().is(':empty')) {
+        $.get(full_path, (content: JQuery.PlainObject) => {
+            const lines = content.split('\n');
+            let lo_line_num = Math.max(line_number - 5, 0);
+            const hi_line_num = Math.min(line_number + 4, lines.length);
+            const rendered_context = lines.slice(lo_line_num, hi_line_num)
+                .map((line: string) => {
+                    line = line.replace(/ /g, '&nbsp;');
+                    line = line.replace(/\t/g, '&nbsp;&nbsp;&nbsp;&nbsp;');
```

Comment:
`&nbsp;` replacements are annoying for copy+paste purposes. Can we use `<pre>` or CSS `white-space: pre`/`pre-wrap`?

Reply:
Yeah. Though copying will still be a bit annoying due to the line numbers.